Correctly handle null array constants from source and metadata

333fred opened the pull request:
Fixes https://github.com/dotnet/roslyn/issues/58598. Fixes https://github.com/dotnet/roslyn/issues/58025. Interpolated string handlers were presuming there would never be a null array in source, and we had a general bug with our string array cracking code that assumed arrays from metadata would also never be null.

333fred commented on src/Compilers/Core/Portable/MetadataReader/PEModule.cs:
```diff
@@ -1903,18 +1904,21 @@ internal static bool CrackStringArrayInAttributeValue(out ImmutableArray<string?
                 uint arrayLen = sig.ReadUInt32();
-                var stringArray = new string?[arrayLen];
-                for (int i = 0; i < arrayLen; i++)
+                if (sig.RemainingBytes >= arrayLen)
```
A null constant in metadata will return `0xFFFFFFFF` for `arrayLen` here. This is the same approach used by the other 2 array crackers in PEModule, `CrackByteArrayInAttributeValue` and `CrackBoolArrayInAttributeValue`.

AlekseyTs replied:
>sig.RemainingBytes >= arrayLen [](http://example.com/codeflow?start=20&length=30)

Since the 0xFFFFFFFF value is part of the specification we probably should have an explicit equality check for it. Same in other similar functions. Also, while in ```CrackByteArrayInAttributeValue``` and ```CrackBoolArrayInAttributeValue``` it is appropriate to compare amount of  bytes to the amount of elements, I don not think it is appropriate here. Even if we have that many bytes, it doesn't mean we have that many strings.

333fred commented:
@dotnet/roslyn-compiler for review.

RikkiGibson commented on src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:
```diff
@@ -6851,6 +6851,36 @@ class C
+
+            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerArgumentAttribute, handler });
+            comp.VerifyDiagnostics(
+                // (4,8): error CS8949: The InterpolatedStringHandlerArgumentAttribute applied to parameter 'CustomHandler' is malformed and cannot be interpreted. Construct an instance of 'CustomHandler' manually.
```
Should we have an error at the declaration site of `C.M` as well?

AlekseyTs commented:
Done with review pass (commit 2)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [1005301270](https://github.com/dotnet/roslyn/pull/58604#issuecomment-1005301270) [](http://example.com/codeflow?ancestors=1005301270,1005301270)

AlekseyTs commented on src/Compilers/Core/Portable/MetadataReader/PEModule.cs:
```diff
@@ -1921,6 +1930,26 @@ internal static bool CrackStringArrayInAttributeValue(out ImmutableArray<string?
+                return false;
+            }
+
+            // II.23.3 says that the length must be a int32 if not null. There can't be negative lengths,
```
>// II.23.3 says that the length must be a int32 if not null [](http://example.com/codeflow?start=12&length=59)

I cannot find where does it say that. Here is what I found: "NumElem is an **unsigned int32** specifying the number of elements"

AlekseyTs commented:
Done with review pass (commit 4)

AlekseyTs commented on src/Compilers/Core/Portable/MetadataReader/PEModule.cs:
```diff
@@ -1921,6 +1930,19 @@ internal static bool CrackStringArrayInAttributeValue(out ImmutableArray<string?
             return false;
         }
 
+        private static bool ValidateMetadataArrayLength(uint length)
```
>ValidateMetadataArrayLength [](http://example.com/codeflow?start=28&length=27)

The name is somewhat confusing, I think. The length isn't invalid, it indicates null value.

AlekseyTs reviewed (approved):
LGTM (commit 5), with rename suggestion.
